source/proximity_list/proximity_list.cpp

```diff
@@ -93,12 +93,20 @@ void ProximityList<T>::build() {
 }
 
 template <typename T, typename Item>
-static auto extract(const Item& item, IndexCoord /*unused*/) -> typename ReturnTypeTrait<T, IndexCoord>::ValueType {
+static auto extract(const Item& item, float /*unused*/, IndexCoord /*unused*/) ->
+    typename ReturnTypeTrait<T, IndexCoord>::ValueType {
     return std::make_pair(item.element, item.coord);
 }
 
 template <typename T, typename Item>
-static auto extract(const Item& item, IndexOnly /*unused*/) -> typename ReturnTypeTrait<T, IndexOnly>::ValueType {
+static auto extract(const Item& item, float distance, IndexCoordDistance /*unused*/) ->
```

Comment:
By templating the 3rd parameter's type, you could only keep only 1 `extract` implementation:
```cpp
 template <typename T, typename IndexType, typename Item,>
 static auto extract(const Item& item, float distance) -> 
    typename ReturnTypeTrait<T, IndexType>::ValueType {
    return std::make_tuple(item.element, item.coord, distance);
}
```
So that later, you only need 
```cpp
 res.push_back(extract<T, Tag>(items[res_ind], distances.at(i)));
```

That should also simplify types to only work on `tuple` (rather than mixing pair, tuple and other)